component/extension.go

```diff
@@ -52,22 +53,39 @@ type ExtensionCreateSettings struct {
 	BuildInfo BuildInfo
 }
 
-// ExtensionFactory is a factory interface for extensions to the service.
-//
-// This interface cannot be directly implemented. Implementations must
-// use the extensionhelper.NewFactory to implement it.
-type ExtensionFactory interface {
-	Factory
+// ExtensionCreateDefaultConfigFunc is the equivalent of component.ExtensionFactory.CreateDefaultConfig()
+type ExtensionCreateDefaultConfigFunc func() config.Extension
+
+// CreateDefaultConfig creates the default configuration for the Extension.
+// This method can be called multiple times depending on the pipeline
+// configuration and should not cause side-effects that prevent the creation
+// of multiple instances of the Extension.
+// The object returned by this method needs to pass the checks implemented by
+// 'configtest.CheckConfigStruct'. It is recommended to have these checks in the
+// tests of any implementation of the Factory interface.
+func (f ExtensionCreateDefaultConfigFunc) CreateDefaultConfig() config.Extension {
+	return f()
+}
+
+// CreateExtensionFunc is the equivalent of component.ExtensionFactory.CreateExtension()
+type CreateExtensionFunc func(context.Context, ExtensionCreateSettings, config.Extension) (Extension, error)
+
+// CreateExtension creates a service extension based on the given config.
+func (f CreateExtensionFunc) CreateExtension(ctx context.Context, set ExtensionCreateSettings, cfg config.Extension) (Extension, error) {
+	return f(ctx, set, cfg)
+}
+
+// ExtensionFactory is a factory for extensions to the service.
+type ExtensionFactory struct {
+	internalinterface.BaseInternal
```

Comment:
Soon we will remove when all factories will be struct and this can be something internal in that package (or removed completely).